File: src/horizon_mcp/tools/profiles.py
"""Certificate profile read-only tools — list and get profiles.

Safety tiers:
    - list_profiles, get_profile: read-only

References:
    - horizon://knowledge/profiles
"""

from __future__ import annotations

import json
import logging
from typing import TYPE_CHECKING, Any

from horizon_mcp.tools._helpers import (
    apply_name_filter,
    build_list_response,
)

if TYPE_CHECKING:
    from mcp.server.fastmcp import FastMCP

logger = logging.getLogger("horizon_mcp.tools.profiles")

_PROFILE_BASE = "/api/v1/certificate/profiles"


def _apply_module_filter(
    items: list[dict[str, Any]], module: str | None,
) -> list[dict[str, Any]]:
    """Client-side filter on the 'module' field (case-insensitive exact match)."""
    if not module:
        return items
    needle = module.lower()
    return [item for item in items if item.get("module", "").lower() == needle]


def register_profile_readonly_tools(mcp: FastMCP) -> None:
    """Register read-only profile tools: list and get (2 tools)."""

    from horizon_mcp.client.state import get_client

    @mcp.tool()
    async def list_profiles(
        max_items: int = 50,
        name_contains: str | None = None,
        module: str | None = None,
    ) -> str:
        """List certificate profiles with optional filtering.

        Safety tier: read-only
        Knowledge: horizon://knowledge/profiles

        Client-side filtering is applied after fetching all profiles from
        the API. Use name_contains for substring search and module for
        exact module type matching.

        Args:
            max_items: Maximum number of profiles to return (default 50).
            name_contains: Case-insensitive substring filter on profile name.
            module: Filter by module type (webra, acme, scep, est, monitored).

        Returns:
            JSON with items, count, total_available, and truncated flag.
        """
        client = get_client()
        data = await client.get(_PROFILE_BASE)
        items: list[dict[str, Any]] = (
            data if isinstance(data, list) else data.get("items", [data])
        )
        items = apply_name_filter(items, name_contains)
        items = _apply_module_filter(items, module)
        return build_list_response(items, max_items, kind="profile")

    @mcp.tool()
    async def get_profile(name: str) -> str:
        """Get full details of a single certificate profile by name.

        Safety tier: read-only
        Knowledge: horizon://knowledge/profiles

        Args:
            name: Exact profile name.

        Returns:
            JSON representation of the profile including all configuration.
        """
        client = get_client()
        result = await client.get(f"{_PROFILE_BASE}/{name}")
        return json.dumps(result)

File: src/tools/profiles.ts
import type { McpServer } from '@modelcontextprotocol/server';
import { z } from 'zod';

import type { HorizonClient } from '../client/http.js';
import { normalizeItems } from './datasources/shared.js';
import {
  applyNameFilter,
  buildListResponse,
  encodePathSegment,
} from './helpers.js';
import { registerTool } from './register.js';

const PROFILE_BASE = '/api/v1/certificate/profiles';

const LIST_PROFILES_CONFIG = {
  description:
    'List certificate profiles with optional filtering.\n\n Ref: horizon://knowledge/profiles.' +
    'Client-side filtering is applied after fetching all profiles. ' +
    'Use name_contains for substring search and module for exact module type matching.',
  inputSchema: z.object({
    max_items: z
      .number()
      .int()
      .positive()
      .max(100)
      .default(50)
      .describe('Maximum number of profiles to return (default 50).'),
    name_contains: z
      .string()
      .optional()
      .describe('Case-insensitive substring filter on profile name.'),
    module: z
      .string()
      .optional()
      .describe('Filter by module type (webra, acme, scep, est, monitored).'),
  }),
};

const GET_PROFILE_CONFIG = {
  description:
    'Get full details of a single certificate profile by name.\n\n Ref: horizon://knowledge/profiles.',
  inputSchema: z.object({
    name: z.string().describe('Exact profile name.'),
  }),
};

function applyModuleFilter(
  items: Record<string, unknown>[],
  module?: string,
): Record<string, unknown>[] {
  if (!module) return items;
  const needle = module.toLowerCase();
  return items.filter((item) => {
    const m = item['module'];
    return typeof m === 'string' && m.toLowerCase() === needle;
  });
}

export function registerProfileTools(
  server: McpServer,
  client: HorizonClient,
): void {
  registerTool(
    server,
    'list_profiles',
    LIST_PROFILES_CONFIG,
    async ({ max_items, name_contains, module }) => {
      const data = await client.get<unknown>(PROFILE_BASE);
      let items = normalizeItems(data);
      items = applyNameFilter(items, name_contains);
      items = applyModuleFilter(items, module);
      return {
        content: [
          {
            type: 'text' as const,
            text: buildListResponse(items, max_items, 'profile'),
          },
        ],
      };
    },
  );

  registerTool(server, 'get_profile', GET_PROFILE_CONFIG, async ({ name }) => {
    const result = await client.get(
      `${PROFILE_BASE}/${encodePathSegment(name)}`,
    );
    return {
      content: [{ type: 'text' as const, text: JSON.stringify(result) }],
    };
  });
}
